engine/consensus/Sealing_Readme.md

```diff
@@ -0,0 +1,89 @@
+
+      Caution: to not compromise sealing liveness, the sealing engine cannot drop any incorporated blocks or any results contained in it. Otherwise, it cannot successfully create a seal for the incorporated results.
+
+- Sealing Engine will forward each receipts and approval to the sealing core. Receipts and approvals are concurrently processed by sealing core.
+- The purpose for the Sealing Core is to
+    1. compute verifier assignments for the incorporated results
+    2. track the approvals that match the assignments
+    3. generate a seal once enough approvals have been collected
+- For sealing core to concurrently process receipts, it maintains an `AssignmentCollectorTree`, which is a fork-aware structure holding the assignments + approvals for each incorporated result. Therefore, it can ingest results and approvals in arbitrary order.
+    - Each vertex (`AssignmentCollector`) in the tree is responsible for collecting approvals one particular incorporated result. Each collector is able to process approvals concurrently. But accessing the tree requires a lock.
+    - A result might be incorporated in multiple blocks on different block forks. Results incorporated on different forks will lead to different assignment for verification nodes. Although the approvals are the same for incorporated results on different fork, a valid approval for a certain incorporated result on a fork might be invalid for another incorporated result in a different fork, because the assignment is different and a verification node is not allowed to verify a result if the result is not assigned to it to verify.
+- The `AssignmentCollectorTree` tracks whether the result is derived from the latest result with a finalized seal. If interim results are unknown (due to concurrent ingestion of incorporated results), approvals are only cached (status `CachingApprovals`). Once all interim results are received, the `AssignmentCollectorTree` changes the status of the result and all its *connected* children to be processable (status `VerifyingApprovals`).
+
+  Furthermore, if results are orphaned because they conflict with a result that has a finalized seal, the processing of all approvals for the conflicting result and any of its derived results is stopped  (status `Orphaned`)
+
+- In order to limit the memory usage of the `AssignmentCollectorTree`, we need to prune it by sealed height. Therefore, we subscribe to finalized block event and use the sealed height to prune the tree.
+
+  When pruning the tree, we keep the node at the same height as the last sealed height, so that we know which fork actually connects to the sealed block
+
+- It means adding a new result to the tree requires a lock, and is a write lock. A write lock won't block reads, so that the worker to process approvals won't be blocked.
+- when a result has been added to a tree, it also means we've received the executed block, because otherwise the block, which includes the receipt, won't pass the validation
```

Comment:
```suggestion
- When a result has been added to a tree, it also means we've received the executed block, because otherwise the block, which includes the receipt, won't pass the validation
```